flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/FailoverRegion.java

```diff
@@ -206,13 +228,15 @@ private void restart(long globalModVersionOfFailover) {
 		try {
 			if (transitionState(JobStatus.CREATED, JobStatus.RUNNING)) {
 				// if we have checkpointed state, reload it into the executions
-				//TODO: checkpoint support restore part ExecutionVertex cp
-				/**
 				if (executionGraph.getCheckpointCoordinator() != null) {
+					// we restart the checkpoint scheduler for
+					// i) enable new checkpoint could be triggered without waiting for last checkpoint expired.
+					// ii) ensure the EXACTLY_ONCE semantics if needed.
+					executionGraph.getCheckpointCoordinator().cancelPendingCheckpoints();
+
 					executionGraph.getCheckpointCoordinator().restoreLatestCheckpointedState(
```

Comment:
I have a suggestion what we should also change after this was merged: inside `restoreLatestCheckpointedState`, the shared state registry is currently always closed and recreated from the checkpoint store (i.e. typically zookeeper). This is not strictly needed, in fact it should only be required if the leader of the job has changed. Doing this for every regional failover is very wasteful. Please take a look at FLINK-6984.

Reply:
Sounds reasonable, and can't believe this JIRA was created almost two years ago!

Reply:
Yes, that sounds crazy but I guess there was always something more important to do and it can also be dangerous to make such changes if the existing code works well.